pkg/volume/portworx/portworx.go

```diff
@@ -0,0 +1,390 @@
+	volume.MetricsProvider
+}
+
+type portworxVolumeMounter struct {
+	*portworxVolume
+	// Filesystem type, optional.
+	fsType string
+	// Specifies whether the disk will be attached as read-only.
+	readOnly bool
+	// diskMounter provides the interface that is used to mount the actual block device.
+	diskMounter *mount.SafeFormatAndMount
+}
+
+var _ volume.Mounter = &portworxVolumeMounter{}
+
+func (b *portworxVolumeMounter) GetAttributes() volume.Attributes {
+	return volume.Attributes{
+		ReadOnly: b.readOnly,
+		Managed:  !b.readOnly,
+		// true ?
+		SupportsSELinux: true,
+	}
+}
+
+// Checks prior to mount operations to verify that the required components (binaries, etc.)
+// to mount the volume are available on the underlying node.
+// If not, it returns an error
+func (b *portworxVolumeMounter) CanMount() error {
+	return nil
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (b *portworxVolumeMounter) SetUp(fsGroup *int64) error {
+	return b.SetUpAt(b.GetPath(), fsGroup)
+}
+
+// SetUpAt attaches the disk and bind mounts to the volume path.
+func (b *portworxVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
+	notMnt, err := b.mounter.IsLikelyNotMountPoint(dir)
+	glog.V(4).Infof("Portworx Volume set up: %s %v %v", dir, !notMnt, err)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("Cannot validate mountpoint: %s", dir)
+		return err
+	}
+	if !notMnt {
+		return nil
+	}
+
+	if _, err := b.manager.AttachVolume(b); err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("Portworx Volume %s attached", b.volumeID)
+
+	if err := os.MkdirAll(dir, 0750); err != nil {
+		return err
+	}
+
+	if err := b.manager.MountVolume(b, dir); err != nil {
+		return err
+	}
+	volume.SetVolumeOwnership(b, fsGroup)
+	glog.V(4).Infof("Portworx Volume %s mounted to %s", b.volumeID, dir)
+	return nil
+}
+
+func (pwx *portworxVolume) GetPath() string {
+	return getPath(pwx.podUID, pwx.volName, pwx.plugin.host)
+}
+
+type portworxVolumeUnmounter struct {
+	*portworxVolume
+}
+
+var _ volume.Unmounter = &portworxVolumeUnmounter{}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (c *portworxVolumeUnmounter) TearDown() error {
+	return c.TearDownAt(c.GetPath())
+}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (c *portworxVolumeUnmounter) TearDownAt(dir string) error {
+	glog.V(4).Infof("Portworx Volume TearDown of %s", dir)
+	// Unmount the bind mount inside the pod
+	util.UnmountPath(dir, c.mounter)
+
+	// Call Portworx Unmount for Portworx's book-keeping.
+	if err := c.manager.UnmountVolume(c, dir); err != nil {
+		return err
+	}
+
+	// Double check if the volume is unmounted
+	notMnt, _ := c.mounter.IsLikelyNotMountPoint(dir)
+	if !notMnt {
+		// Mountpoint still exists
+		return fmt.Errorf("Failed to unmount Portworx Volume at %s", dir)
+	}
+
```

Comment:
yes that is also been done in util.UnmountPath().
Will remove this too.
So it will look like this

1.util.UnmountPath()
2. call portworx.UnmountVolume
3. call portworx.DetachVolume